File: App/Features/LiveMatch/LiveShareCoordinator.swift
import Catalog
import Domain
import Foundation
import Observation
import Store
import SwiftData
import Sync

/// Doc 09 « Fin de partie » (révisé) — pendant de `MatchConnectionCoordinator` côté hôte : porte
/// la session de partage (`LiveSession`/`SupabaseTransport`/code d'appairage) pour toute la durée
/// de l'app, plutôt que pour la durée d'un `LiveMatchModel`/`LiveMatchView`, qui se recrée à
/// chaque nouvelle partie. C'est ce découplage qui permet à une session de survivre à la fin
/// d'une partie et d'enchaîner sur la suivante — même jeu rejoué ou jeu différent — jusqu'à un
/// arrêt explicite (« Arrêter le partage »), qui reste la seule façon de la terminer.
@MainActor
@Observable
final class LiveShareCoordinator {
  static let shared = LiveShareCoordinator()

  private let catalog = GameCatalog.embedded

  private var session: LiveSession?
  private var transport: SupabaseTransport?
  private var match: MatchRecord?
  private var repository: MatchRepository?
  private var sharingTasks: [Task<Void, Never>] = []

  private(set) var pairingCode: String?
  private(set) var connectedPeers: [LiveSession.ConnectedPeer] = []
  /// La partie actuellement diffusée aux pairs connectés — `nil` tant qu'aucune session n'est
  /// active. `LiveMatchModel` compare son propre `match.id` à celui-ci pour savoir s'il est,
  /// lui, la partie affichée par la session en cours (`isSharing`).
  private(set) var attachedMatchID: UUID?
  /// Doc 09 « Fin de partie » — identifiant stable de la session, indépendant du `matchID`
  /// courant. `LiveMatchModel` s'en sert pour donner à `MatchLiveActivityController` une clé
  /// d'Activity qui survit à un changement de partie (voir `MatchLiveActivityController.refresh`).
  private(set) var sessionID: UUID?
  private(set) var allowsContributors = true
  var isSharing: Bool { session != nil }

  /// Doc utilisateur — remontée : ouvrir l'écran d'une partie substituait silencieusement ce
  /// que voient les pairs connectés dès qu'une session diffusait déjà une *autre* partie encore
  /// en cours. `LiveMatchModel` s'en sert pour savoir si un tel rattachement doit être proposé
  /// tel quel (enchaînement voulu, doc 09 « Fin de partie ») ou demander confirmation d'abord
  /// (deux parties bien distinctes, encore en cours toutes les deux).
  var attachedMatchIsConcluded: Bool {
    guard let match else { return true }
    return match.status == .ended || match.status == .abandoned
  }

  /// Nom du jeu actuellement diffusé — pour le message de confirmation avant de le remplacer.
  var attachedGameName: String? {
    guard let match else { return nil }
    return (try? catalog.definition(for: match.gameID, version: match.rulesVersion))?.name.fr
      ?? match.gameID
  }

  /// Doc utilisateur — un `LiveMatchModel` se recrée à chaque partie ; les flux de `LiveSession`
  /// sont documentés à usage unique (voir `SharedMatchModel`, bug BLE-era d'un second abonné
  /// privé du flux), donc ce coordinateur en reste l'unique consommateur pour toute la durée de
  /// la session et republie ce qu'il faut savoir via ces propriétés `@Observable`, lisibles par
  /// n'importe quel `LiveMatchModel` actif (`.onChange(of:)`) sans jamais re-souscrire
  /// directement à `session.events`.
  private(set) var remoteEventToken = UUID()
  private(set) var remoteEventMatchID: UUID?
  private(set) var remoteEventDeviceID: String?
  private(set) var remoteEventIsRoundCommit = false

  private init() {}

  /// Démarre une toute nouvelle session si aucune n'est active ; sinon équivalent à
  /// `attach(match:context:)` — permet à `ShareSessionView` d'appeler la même méthode dans les
  /// deux cas sans avoir à distinguer « première partie partagée » de « partie suivante ».
  func startSharing(
    match: MatchRecord, context: ModelContext, deviceName: String, allowsContributors: Bool
  ) async throws {
    guard session == nil else {
      await attach(match: match, context: context)
      return
    }

    let repository = MatchRepository(context: context)
    let newSessionID = UUID()
    let newSession = LiveSession(deviceID: DeviceIdentity.current, catalog: catalog)
    let code = LiveSession.generatePairingCode()
    try await newSession.startHosting(
      log: try repository.currentLog(for: match),
      sessionID: newSessionID,
      pairingCode: code,
      allowsContributors: allowsContributors
    )

    let newTransport = SupabaseTransport(deviceID: DeviceIdentity.current, deviceName: deviceName)
    try await newTransport.advertise(
      sessionID: newSessionID,
      matchID: match.id,
      gameID: match.gameID,
      participantCount: match.participants.count,
      pairingCode: code
    )

    session = newSession
    transport = newTransport
    self.repository = repository
    self.match = match
    pairingCode = code
    self.allowsContributors = allowsContributors
    attachedMatchID = match.id
    sessionID = newSessionID
    connectedPeers = []

    let acceptTask = Task { [weak self] in
      for await incoming in newTransport.acceptIncoming() {
        guard self != nil else { return }
        await newSession.acceptConnection(incoming)
      }
    }
    let eventsTask = Task { [weak self] in
      for await stamped in newSession.events {
        self?.handleRemoteEvent(stamped)
      }
    }
    let peersTask = Task { [weak self] in
      for await peers in newSession.peerUpdates {
        self?.connectedPeers = peers
      }
    }
    sharingTasks = [acceptTask, eventsTask, peersTask]
  }

  /// Doc 09 « Fin de partie » — appelé pour chaque `LiveMatchModel` créé (nouvelle partie ou
  /// reprise d'une partie en cours) : silencieusement sans effet si aucune session n'est active,
  /// et sans effet si déjà attaché à cette partie. Sinon pousse son journal complet à la session
  /// déjà ouverte (`LiveSession.switchMatch`) — sans rouvrir le canal, changer la clé, ni
  /// déconnecter les pairs déjà présents. C'est ce seul appel, fait depuis l'initialiseur de
  /// `LiveMatchModel`, qui fait qu'une nouvelle partie rejoint automatiquement une session déjà
  /// active, sans repasser par « Partager en direct ».
  func attach(match: MatchRecord, context: ModelContext) async {
    guard let session else { return }
    guard attachedMatchID != match.id else { return }

    let repository = MatchRepository(context: context)
    guard let log = try? repository.currentLog(for: match) else { return }
    guard (try? await session.switchMatch(log: log)) != nil else { return }
    try? await transport?.updateActiveMatch(
      matchID: match.id, gameID: match.gameID, participantCount: match.participants.count)

    self.repository = repository
    self.match = match
    attachedMatchID = match.id
  }

  /// Doc utilisateur — appelé après chaque écriture locale de l'hôte sur la partie actuellement
  /// attachée (saisie, annulation, fin de partie) : ne fait plus qu'une resynchronisation, ne
  /// termine plus jamais la session à elle seule (doc 09 « Fin de partie », révisé — l'ancien
  /// comportement arrêtait automatiquement le partage à la conclusion de la partie).
  func syncLog(for matchID: UUID) async {
    guard let session, let match, let repository, attachedMatchID == matchID else { return }
    guard let log = try? repository.currentLog(for: match) else { return }
    try? await session.syncHostLog(log)
  }

  /// Doc utilisateur P9 — s'applique aux prochaines connexions, pas aux contributeurs déjà
  /// connectés (voir `LiveSession.setAllowsContributors`).
  func setAllowsContributors(_ allowed: Bool) async {
    allowsContributors = allowed
    await session?.setAllowsContributors(allowed)
  }

  /// Seul point d'arrêt d'une session (doc 09 « Fin de partie ») — reprend telle quelle la
  /// logique de fermeture qui vivait auparavant dans `LiveMatchModel.stopSharing`.
  func stopSharing() async {
    await session?.stopHosting()
    for task in sharingTasks { task.cancel() }
    sharingTasks = []
    await transport?.stopAdvertising()
    transport = nil
    session = nil
    repository = nil
    match = nil
    pairingCode = nil
    connectedPeers = []
    attachedMatchID = nil
    sessionID = nil
    allowsContributors = true
  }

  /// Persiste une manche acceptée d'un contributeur distant (`LiveSession.events` ne porte
  /// jamais les écritures locales de l'hôte — voir `LiveSession.hostCommit`, jamais appelé par
  /// `session.propose` côté hôte) et republie ce qu'il faut savoir pour que le `LiveMatchModel`
  /// concerné se resynchronise (`refreshFromRemote`).
  private func handleRemoteEvent(_ stamped: StampedEvent) {
    guard let match, let repository else { return }
    guard (try? repository.appendRemoteEvent(stamped, to: match, catalog: catalog)) != nil else {
      return
    }

    remoteEventMatchID = match.id
    remoteEventDeviceID = stamped.deviceID
    if case .roundCommitted = stamped.event {
      remoteEventIsRoundCommit = true
    } else {
      remoteEventIsRoundCommit = false
    }
    remoteEventToken = UUID()
  }
}

File: android/app/src/main/kotlin/com/cacompte/app/livesync/LiveShareCoordinator.kt
package com.cacompte.app.livesync

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.cacompte.domain.engine.MatchEvent
import com.cacompte.domain.engine.StampedEvent
import com.cacompte.domain.rules.GameCatalog
import com.cacompte.store.MatchEntity
import com.cacompte.store.MatchRepository
import com.cacompte.sync.LiveSession
import com.cacompte.sync.SupabaseTransport
import com.cacompte.sync.WireMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.launch
import java.util.UUID

/** Doc 09 — notifie [com.cacompte.app.features.livematch.LiveMatchViewModel] qu'une manche
 * distante (contributeur) vient d'être acceptée et persistée pour `matchID`, afin qu'il recharge
 * son état et affiche éventuellement un bandeau d'activité. Miroir des propriétés
 * `remoteEventMatchID`/`remoteEventDeviceID`/`remoteEventIsRoundCommit` de
 * `LiveShareCoordinator.swift`, réunies dans un seul événement plutôt qu'un jeton `.onChange` —
 * idiome SwiftUI sans équivalent direct en Compose. */
data class RemoteMatchUpdate(
    val matchID: UUID,
    val deviceName: String?,
    val isRoundCommit: Boolean,
)

/**
 * Coordinateur **hôte**, durée de vie de l'application (une seule instance dans
 * [com.cacompte.app.di.AppContainer]) — miroir de `LiveShareCoordinator.swift`. Possède le
 * `LiveSession`/`SupabaseTransport` de la session partagée courante indépendamment de l'écran
 * `LiveMatchScreen` affiché : la partie reste partagée même si l'utilisateur navigue ailleurs.
 *
 * Simplification assumée par rapport à Apple : pas d'enchaînement d'une session sur plusieurs
 * parties successives (`attach`/`switchMatch`) — démarrer un nouveau partage ferme d'abord
 * l'ancien. Suffisant pour partager une partie à la fois ; l'enchaînement pourra être ajouté
 * plus tard si le besoin se confirme.
 */
class LiveShareCoordinator(
    private val catalog: GameCatalog,
    private val matchRepository: MatchRepository,
    private val resolveDeviceID: suspend () -> String,
    private val scope: CoroutineScope,
) {
    private var session: LiveSession? = null
    private var transport: SupabaseTransport? = null
    private var attachedMatch: MatchEntity? = null
    private var backgroundJobs = mutableListOf<Job>()

    var pairingCode: String? by mutableStateOf(null)
        private set
    var connectedPeers: List<LiveSession.ConnectedPeer> by mutableStateOf(emptyList())
        private set
    var attachedMatchID: UUID? by mutableStateOf(null)
        private set
    var allowsContributors: Boolean by mutableStateOf(true)
        private set

    private val remoteMatchUpdatesFlow = MutableSharedFlow<RemoteMatchUpdate>(extraBufferCapacity = 8)
    val remoteMatchUpdates: SharedFlow<RemoteMatchUpdate> = remoteMatchUpdatesFlow

    /** Démarre le partage de [match] — ou ne fait rien si elle est déjà la partie partagée. */
    suspend fun startSharing(
        match: MatchEntity,
        participantCount: Int,
        deviceName: String,
        allowsContributors: Boolean,
    ) {
        if (attachedMatchID == match.id) return
        if (attachedMatchID != null) stopSharing()

        val deviceID = resolveDeviceID()
        val newSession = LiveSession(deviceID = deviceID, catalog = catalog, scope = scope)
        val newTransport =
            SupabaseTransport(
                deviceID = deviceID,
                deviceName = deviceName,
                scope = scope,
                platform = WireMessage.Platform.Android,
            )

        val sessionID = UUID.randomUUID()
        val code = LiveSession.generatePairingCode()
        newSession.startHosting(
            initialLog = matchRepository.currentLog(match),
            sessionID = sessionID,
            pairingCode = code,
            allowsContributors = allowsContributors,
        )
        newTransport.advertise(
            sessionID = sessionID,
            matchID = match.id,
            gameID = match.gameID,
            participantCount = participantCount,
            pairingCode = code,
        )

        session = newSession
        transport = newTransport
        attachedMatch = match
        attachedMatchID = match.id
        pairingCode = code
        this.allowsContributors = allowsContributors
        connectedPeers = emptyList()

        backgroundJobs +=
            scope.launch {
                newTransport.acceptIncoming().collect { transportSession ->
                    newSession.acceptConnection(transportSession)
                }
            }
        backgroundJobs += scope.launch { newSession.events.collect(::handleRemoteEvent) }
        backgroundJobs += scope.launch { newSession.peerUpdates.collect { peers -> connectedPeers = peers } }
    }

    /** Doc « Fin de partie » — republie le journal complet vers les pairs déjà connectés après
     * chaque écriture hôte locale (`LiveSession.syncHostLog` ne diffuse que les nouveaux
     * événements). No-op si [matchID] n'est pas la partie actuellement partagée. */
    suspend fun syncLog(matchID: UUID) {
        val match = attachedMatch ?: return
        if (match.id != matchID) return
        val current = requireNotNull(matchRepository.match(matchID))
        attachedMatch = current
        session?.syncHostLog(matchRepository.currentLog(current))
    }

    suspend fun setAllowsContributors(allowed: Boolean) {
        session?.setAllowsContributors(allowed) ?: return
        allowsContributors = allowed
    }

    suspend fun stopSharing() {
        for (job in backgroundJobs) job.cancel()
        backgroundJobs.clear()
        session?.stopHosting()
        transport?.stopAdvertising()
        session = null
        transport = null
        attachedMatch = null
        attachedMatchID = null
        pairingCode = null
        connectedPeers = emptyList()
    }

    private suspend fun handleRemoteEvent(stamped: StampedEvent) {
        val match = attachedMatch ?: return
        try {
            matchRepository.appendRemoteEvent(stamped, match, catalog)
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (error: Exception) {
            return
        }
        attachedMatch = requireNotNull(matchRepository.match(match.id))
        val deviceName = connectedPeers.firstOrNull { it.deviceID == stamped.deviceID }?.deviceName
        remoteMatchUpdatesFlow.emit(
            RemoteMatchUpdate(
                matchID = match.id,
                deviceName = deviceName,
                isRoundCommit = stamped.event is MatchEvent.RoundCommitted,
            ),
        )
    }
}
